### ISSUE TITLE
Adaptive integer aspect ratio correction.

### ISSUE THREAD

**Issue by PorscheMonty** [chocolate-doom/chocolate-doom#967]
Something interesting was recently brought to my attention.

Chocolate Doom can apply aspect ratio correction so that the output image will always be that of the ideal 4:3, but unless the display area is at least 1600x1200, this inevitably introduces scaling error.

A way of performing aspect ratio correction without introducing scaling error would be to stick to integer values when altering both horizontal and vertical resolutions independently, and the higher the native resolution, the closer to the aforementioned ideal ratio it will be. E.g, full HD would allow for the accommodation of a 4:5 (1280x1000-1.28) upscaled image which's barely indistinguishable from the target 1600x1200-1.33.

Perhaps this aspect ratio correction method could be implemented as an alternative that could be enabled from "aspect_ratio_correct" with a "2" while possibly also allowing the user to manually enter a custom value (H:V).

Now that I'm temporarily left without a functioning CRT (caps in the neck board gone, and nasty desoldering job ahead just to remove the enclosure), I've been thinking more seriously about switching to one of these new flatpanel technologies currently in the horizon just to alleviate the issues I now face with this cheap, spare LCD that\s hardly even acceptable for office work.

Thanks for your time and sorry for the inconveniences.

**Comment by fragglet** [comment]
Apart from 1600x1200 (or integer multiples) it's not really possible to do both 4:3 aspect ratio correction and exact integer scale-up.

What you can get is integer in one dimension. For example 800x600 is an exact integer multiple vertically but not horizontally. If the blurriness is only in one dimension then it's not quite as bad. We have a `integer_scaling` config file option now that ensures you're integer scaled in at least one dimension.

What you're advocating is to instead sacrifice the correct ratio (ie. 1.28 is pretty close to 1.33). It's an interesting idea, but it's worth noting that we did previously remove the 1280x1000 software screen mode in d91383ee4057a2fc6fc2fb80144abb34273a6d7f because it seemed redundant.

**Comment by PorscheMonty** [comment]
Well, I was thinking more in the terms of auto-selecting an aspect ratio-corrected integer output based on the display mode of choice (native). There are but 4 integer-scaling options that make sense (horizontal resolution still larger than vertical), yet only 1 is ideal (*) and the other 3 compromises:

http://oi64.tinypic.com/efgx1v.jpg

640x600 = 2:3 (1.06)
960x800 = 3:4 (1.20)
1280x1000 = 4:5 (1.28)
1600x1200 = 5:6 (1.33) *

I myself would be stuck with 1.20, which isn't as bad as it sounds.

Now you never know what kind of hardware's going to be used to display Chocolate Doom (given it can be ported to virtually anything) so it is possible that, for example, only an 800x600 display may be available, forcing the user to choose between 640x400 or 640x600 if scaling error is absolutely and categorically unacceptable for one reason or another.

Should this be at all considered, it would obviously have to be optional, defaulting always to the current method of correction which's what most people will be comfortable with.

**Comment by linguica** [comment]
Is there some philosophical reason why Chocolate Doom doesn’t support a -geom XxY command line parameter that would allow arbitrarily choosing a window resolution and thereby aspect ratio?

**Comment by chungy** [comment]
@linguica it already has that feature (see `-geometry WxH`)

**Comment by linguica** [comment]
Yeah, I feel like allowing arbitrary dimensions for -geometry would solve everyone's problems, right?
